pkg/awsapi/iam.go

```diff
@@ -63,53 +79,16 @@ func NewIAM(region string) *IAM {
 	}
 }
 
-// CreateRole creates/updates the role
-func (i *IAM) CreateRole(ctx context.Context, req IAMRoleRequest) (*IAMRoleResponse, error) {
-	log := log.Logger(ctx, "awsapi", "iam", "CreateRole")
+// EnsureRole ensures that a role exists, and that it has the appropriate configuration
+func (i *IAM) EnsureRole(ctx context.Context, req IAMRoleRequest) (*IAMRoleResponse, error) {
+	log := log.Logger(ctx, "awsapi", "iam", "EnsureRole")
 	log = log.WithValues("roleName", req.Name)
-	log.V(1).Info("Initiating api call")
-	input := &iam.CreateRoleInput{
-		AssumeRolePolicyDocument: aws.String(req.TrustPolicy),
-		RoleName:                 aws.String(req.Name),
-		Description:              aws.String(req.Description),
-		MaxSessionDuration:       aws.Int64(req.SessionDuration),
-		PermissionsBoundary:      aws.String(req.ManagedPermissionBoundaryPolicy),
-	}
 
-	if err := input.Validate(); err != nil {
-		log.Error(err, "input validation failed")
-		return nil, err
-	}
-
-	roleAlreadyExists := false
-	iResp, err := i.Client.CreateRole(input)
+	// Get the role, or create it.
+	log.V(1).Info("Verifying that IAM Role exists")
+	role, err := i.GetOrCreateRole(ctx, req)
 	if err != nil {
-		if aerr, ok := err.(awserr.Error); ok {
-			switch aerr.Code() {
-			// Update the role to the latest spec if it is existed already
-			case iam.ErrCodeEntityAlreadyExistsException:
-				roleAlreadyExists = true
-				log.Info(iam.ErrCodeEntityAlreadyExistsException)
-			case iam.ErrCodeLimitExceededException:
-				log.Error(err, iam.ErrCodeLimitExceededException)
-			case iam.ErrCodeNoSuchEntityException:
-				log.Error(err, iam.ErrCodeNoSuchEntityException)
-			case iam.ErrCodeServiceFailureException:
-				log.Error(err, iam.ErrCodeServiceFailureException)
-			default:
-				log.Error(err, aerr.Error())
-			}
-		}
-		if !roleAlreadyExists {
-			return nil, err
-		}
-	}
-
-	resp := &IAMRoleResponse{}
-
-	if !roleAlreadyExists {
```

Comment:
So if an exception for `ErrCodeEntityAlreadyExistsException` comes back, then this is set `True`. If it is set to `True`, the code never sets the `resp.RoleARN` or `resp.RoleID` fields in the `IAMRoleResponse` object, which is what triggers the ServiceAccount annotation to get set to an empty string. In the event that the role already exists, we want the code to be intelligent and get the ARN/ID for us.